File: Python/Lacewing.py
from datetime import datetime
import re

from output_log import output_txt_path

class Lacewing:
    
    def __init__(self):
        ## General info
        self.device = None
        self.ROWS = 78
        self.COLS = 56
        self.port = None  # the COM port to which the device is connected
        
        ## Clocks
        self.Clock = None
        
        try:
            import importlib
            lacewing_cmd = importlib.import_module('Lacewing_Cmd_Chiara')
            self.device = lacewing_cmd.Debug_Command  # creates the object
        except ImportError:
            print("Warning: Lacewing_Cmd_Chiara module not found. Some functionality may be limited.")
            
    ## Destructor
    def Disconnect(self):
        if self.device:
            self.device.close_serial()
        self.Clock = datetime.now()
        with open(output_txt_path(), "a") as OUTPUT:
            comm = f"Serial port {self.port} closed. Device disconnected on {self.Clock.strftime('%d/%m/%Y at %H:%M:%S')}"
            OUTPUT.write(comm + "\n")
            print(comm)
        self.port = None
        
    ## Connect
    def Connect(self, port):
        self.port = port
        if self.device:
            self.device.open_serial(port)  # it connects to the device
            self.device.set_timeout(1000000000)  # i need to extend the timeout
            self.device.execute_cmd('ttn_init 3 50')  # initialise the device
            self.Clock = datetime.now()
            with open(output_txt_path(), "a") as OUTPUT:
                comm = f"Serial port {port} opened. Device connected and initialised on {self.Clock.strftime('%d/%m/%Y at %H:%M:%S')}"
                OUTPUT.write(comm + "\n")
                print(comm)
                
    ## FindInfo
    def FindInfo(self):
        if not self.device:
            return [], []
        info = self.device.list_serial()  # list the serial port connected to the pc
        # it converts the py.entry in a list of strings
        name = [str(item) for item in info[0]]
        port = [str(item) for item in info[1]]
        return name, port
        
    ## CheckChip
    def CheckChip(self):
        if not self.device:
            return
        r = self.device.execute_cmd('ttn_check_status')  # 0 not available, 1 electrically active, 2 chemically active, 3 both
        with open(output_txt_path(), "a") as OUTPUT:
            if r == 3:
                comm = 'Chip ready'
            elif r == 2:
                comm = 'Chip not chemically active'
            elif r == 1:
                comm = 'Chip not electrically active'
            elif r == 0:
                comm = 'Chip not available'
            
            OUTPUT.write(comm + "\n")
            print(comm)
        return r
            
    ## Calibration
    def Calibration(self):
        if not self.device:
            return
        Vref = self.device.execute_cmd('ttn_sweep_search_vref')
        Vref_V = (Vref * 10 / 4095) - 5  # theoretical calculation, the real voltage can be slightly different
        with open(output_txt_path(), "a") as OUTPUT:
            comm = f"Chip is calibrated. Vref is {Vref_V} V"
            OUTPUT.write(comm + "\n")
            print(comm)
        return Vref_V
        
    ## PixelStatus
    def PixelStatus(self):
        if not self.device:
            return
        a = self.device.execute_cmd('ttn_eval_pixel')  # check status of each pixels (511 active, o discharge too fast, 1023 discharge too slow)
        array_status = [float(item) for item in a]
        return array_status
        
    ## Calibrated Array
    def CalibArray(self):
        if not self.device:
            return
        self.device.execute_cmd('ttn_temp_init')
        a = self.device.execute_cmd('ttn_cali_vs')
        array_calibrated = [float(item) for item in a]
        return array_calibrated

File: Matlab/Lacewing.m
classdef Lacewing < handle
    properties (GetAccess = 'public', SetAccess = 'public', SetObservable)
        device = [];
        ROWS = 78;
        COLS = 56;
        port = []; %the COM port to which the device is connected
        

        % Clocks
        Clock
    end

    events
    end

     methods(Access = public)

         %% Constructor
        function obj = Lacewing
            py.importlib.import_module('Lacewing_Cmd_Chiara');
            obj.device=py.Lacewing_Cmd_Chiara.Debug_Command; % creates the object             
        end

        %% Distructor
        function Disconnect(obj)
            obj.device.close_serial
            diary on
            comm =['Serial port ', convertStringsToChars(obj.port), ' closed. Device disconnected on ' , num2str(obj.Clock(3)), '/' , num2str(obj.Clock(2)), '/' , num2str(obj.Clock(1)), ' at ', num2str(obj.Clock(4)) , ':' , num2str(obj.Clock(5)) ,':' ,num2str(obj.Clock(6))];
            disp(comm)
            diary off
            obj.port = [];
        end


        %% Connect
        function Connect(obj,port)
            obj.port = port;
            obj.device.open_serial(port); %it connects to the device
            obj.device.set_timeout(1000000000); % i need to extend the timeout
            obj.device.execute_cmd('ttn_init 3 50'); %initailise th device 
            obj.Clock = clock;
            diary OUTPUT
            comm =['Serial port ', convertStringsToChars(obj.port), ' opened. Device connected and initialised on ' , num2str(obj.Clock(3)), '/' , num2str(obj.Clock(2)), '/' , num2str(obj.Clock(1)), ' at ', num2str(obj.Clock(4)) , ':' , num2str(obj.Clock(5)) ,':' ,num2str(obj.Clock(6))];
            disp(comm)
            diary off
        end

        %% FindInfo
        function [name,port] = FindInfo(obj)
            info=obj.device.list_serial; %list the serial port connected to the pc
            name = cellfun(@string,cell(info{1})); %it converts the py.entry in a string
            port = cellfun(@string,cell(info{2})); %it converts the py.entry in a string
        end

        %% CheckChip
        function r=CheckChip(obj)
            r=obj.device.execute_cmd('ttn_check_status'); % 0 not available, 1 electrically active, 2, chemically active,  3 both
            if r == 3
                diary on
                disp ('Chip ready')
                diary off
            elseif r == 2
                diary on
                disp ('Chip not chemically active')
                diary off
            elseif r == 1
                diary on 
                disp ('Chip not electrically active')
                diary off
            elseif r == 0
                diary on
                disp('Chip not available')
                diary off
            end
        end

        %% RefCalibration
        function Vref_V = Calibration(obj)
            Vref=obj.device.execute_cmd('ttn_sweep_search_vref');
            Vref_V = (Vref * 10 / 4095) - 5;  % heoritial calculation, the real voltage can be slightly different
            diary on
            disp (['Chip is calibrated. Vref is ' , num2str(Vref_V), ' V'])
            diary off
        end

        %% PixelStatus

        function array_status = PixelStatus(obj)
            a = obj.device.execute_cmd('ttn_eval_pixel'); % check status of each pixels (511 active, o discharge too fast, 1023 doscharge too slow)
            array_status = cellfun(@double,cell(a));
        end

        %% Calibrated Array

        function array_calibrated = CalibArray(obj)
            obj.device.execute_cmd('ttn_temp_init');
            a = obj.device.execute_cmd('ttn_cali_vs');
            array_calibrated = cellfun(@double,cell(a));
%             figure(2)
%             surf(flipud(reshape(array_calibrated,obj.ROWS,obj.COLS))); view(2);
%             axis tight
%             title('Array calibrated');
%             xlabel('COLS')
%             ylabel('ROWS')
        end

        %% RunOut % modify to add listner for stop and add plot in the window interface

%         function RunOut(obj,time,time_unit)
%             if time_unit == 'sec'
%                 t=time;
%             elseif time_unit == 'min'
%                 t=time*60;
%             elseif time_unit == 'hour'
%                 t=time*60*60;
%             elseif time_unit == 'day'
%                 t=time*60*60*24;                
%             end
%             obj.device.execute_cmd('ttn_temp_init');
%             RunWait(obj,t)
%         end
% 
%         function RunWait(obj,time_sec)
%             t_stamp=0;
%             t=[];
%             avg_px=[];
%             tstart=tic;
%             while t_stamp < time_sec
%                 readout=obj.device.execute_cmd('ttn_readout_vs');
%                 if ~isempty(readout)
%                     t_stamp=toc(tstart);
%                     t=[t t_stamp];
%                     c_readout=cell(readout);
%                     A_readout = cellfun(@double,c_readout);
%                     figure(3)
%                     surf(flipud(reshape(A_readout,obj.ROWS,obj.COLS))); view(2);
%                     axis tight
%                     title(['Array at timebut I frame t = ',num2str(t_stamp), ' sec']);
%                     xlabel('COLS')
%                     ylabel('ROWS')
%                     figure(5)
%                     avg_px=[avg_px mean(A_readout)];
%                     plot(t,avg_px)
%                 end
%                 readout=[];  
%             end
%         end



    
     end

end